ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/OngoingAuthorizationObserver.scala

```diff
@@ -3,14 +3,78 @@
 
 package com.daml.ledger.api.auth
 
+import java.time.Instant
+
+import akka.actor.{Cancellable, Scheduler}
+import com.daml.error.DamlContextualizedErrorLogger
+import com.daml.error.definitions.LedgerApiErrors
+import com.daml.ledger.api.auth.interceptor.AuthorizationInterceptor
+import com.daml.ledger.participant.state.index.v2.UserManagementStore
+import com.daml.lf.data.Ref
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.platform.server.api.validation.ErrorFactories
+import io.grpc.StatusRuntimeException
 import io.grpc.stub.ServerCallStreamObserver
 
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+import scala.util.{Failure, Success}
+
+/** @param userRightsCheckIntervalInSeconds - determines the interval at which to check whether user rights state has changed.
+  *                                         Also, double of this value serves as timeout value for subsequent user rights state checks.
+  */
 private[auth] final class OngoingAuthorizationObserver[A](
     observer: ServerCallStreamObserver[A],
-    claims: ClaimSet.Claims,
-    authorized: ClaimSet.Claims => Either[AuthorizationError, Unit],
-    throwOnFailure: AuthorizationError => Throwable,
-) extends ServerCallStreamObserver[A] {
+    originalClaims: ClaimSet.Claims,
+    nowF: () => Instant,
+    errorFactories: ErrorFactories,
+    userManagementStore: UserManagementStore,
+    implicit val ec: ExecutionContext,
+    userRightsCheckIntervalInSeconds: Int,
+    akkaScheduler: Scheduler,
+)(implicit loggingContext: LoggingContext)
+    extends ServerCallStreamObserver[A] {
+  self =>
+
+  private val logger = ContextualizedLogger.get(getClass)
+  private val errorLogger = new DamlContextualizedErrorLogger(logger, loggingContext, None)
+
+  @volatile private var lastUserRightsCheckTime = nowF()
+
+  private lazy val userId = originalClaims.applicationId.fold[Ref.UserId](
+    throw new RuntimeException(
+      "Claims were resolved from a user but userId (applicationId) is missing in the claims."
+    )
+  )(Ref.UserId.assertFromString)
+
+  private val cancellableO: Option[Cancellable] = {
+    if (originalClaims.resolvedFromUser) {
+      val delay = userRightsCheckIntervalInSeconds.seconds
+      // Note: https://doc.akka.io/docs/akka/2.6.13/scheduler.html states that:
+      // "All scheduled task will be executed when the ActorSystem is terminated, i.e. the task may execute before its timeout."
+      val c = akkaScheduler.scheduleWithFixedDelay(initialDelay = delay, delay = delay)(runnable =
```

Comment:
Akka scheduler is a good approach, if you need it scalable, and don't care much about precision (I think this is the case here)